distribution/client/src/lib/update.ts

```diff
@@ -163,57 +198,76 @@ export default class Update {
-              'to life. Please report this issue to the Jenkins Evergreen team. ' +
-              'Do not shutdown your instance as we have been notified of this failure ' +
-              'and are trying to understand what went wrong to push a new update that ' +
-              'will fix things.';
-            logger.error(failedToRollbackMessage);
-            UI.publish(failedToRollbackMessage);
-
-            // Not throwing an Error here as we want the client to keep running and ready
-            // since the next available UL _might_ fix the issue
-          } else {
-
-            const errorMessage = `Jenkins detected as unhealthy. Rolling back to previous update level (${healthState.message}).`;
-            UI.publish(errorMessage);
-            logger.warn(errorMessage);
-
-            this.snapshotter.revertToLevelBefore(this.getCurrentLevel());
-            this.revertToPreviousUpdateLevel();
-            this.restartJenkins(true);
-          }
-        }
+    return jenkinsIsRestarting
+      .then(() => this.healthChecker.check())
+      .then(() => {
+        logger.info('Jenkins healthcheck after restart succeeded! Yey.');
+        return true;
+
+      }).catch((error) => { // first catch, try rolling back
+
+        const errorMessage = `Jenkins detected as unhealthy. Rolling back to previous update level (${error}).`;
+        logger.warn(errorMessage);
+        UI.publish(errorMessage);
+
+        return this.revertToPreviousUpdateLevel(); // FIXME: async issue!
+
+      }).catch((error) => { // second time wrong, stop trying and just holler for help
+
+        // Quick notice sketch, but I do think we need a very complete and informative message
+        const failedToRollbackMessage =
+          'Ooh noes :-(. We are terribly sorry but it looks like Jenkins failed to ' +
+          'upgrade, but even after the automated rollback we were unable to bring ' +
+          'to life. Please report this issue to the Jenkins Evergreen team. ' +
+          'Do not shutdown your instance as we have been notified of this failure ' +
+          'and are trying to understand what went wrong to push a new update that ' +
+          'will fix things.';
+        logger.error(failedToRollbackMessage, error);
+        UI.publish(failedToRollbackMessage);
+
+        // Not throwing an Error here as we want the client to keep running and ready
+        // since the next available UL _might_ fix the issue
+
+      }).finally(() => {
         Storage.removeBootingFlag();
-      }).catch((errors) => {
-        logger.warn(`TODO ${errors}`);
-      }); // TODO catch?
+        return true;
+      });
+
   }
 
+  /*
+   * 1) Taint current level
+   * 2) query and trigger a new update
+   *    (failing UL has been marked tainted for the instance, so should be served the/a previous UL)
+   * FIXME: 'the/a previous UL' => current code will serve "a" UL. We actually want "the"
```

Comment:
This FIXME comment I am failing to parse, what's the thing to fix here?

Reply:
Because of how we handle updates query, imagine the following case:

* an instance is on UL50
* for some reason, this UL is globally tainted after the fact
* a new UL51 is pushed
* the instance updates to UL51, but fails, so initiate a rollback
* it will per-instance taint UL51, then asks for where to go
* the server will tell the instance to go to UL49, not UL50. So technically, this could cause issues, especially with the data snapshoting system and the associated restore...

For now, I've ignored this issue, and granted this should almost never actually happen given host fast updates are pushed. But it's algorithmically and generally still _possible_.

So with the, granted, stupid phrasing above with _a/the_, I mean we normally expect THE previous UL (UL50 above), not just ANother UL (UL49 in the example).

Is this clearer?

Reply:
Yes, clearer, please put this in the comments.

Reply:
done